File: packages/neovim-lean-ctx/lua/lean-ctx/binary.lua
local M = {}

local cached_path = nil

local candidates = {
  "lean-ctx",
  vim.fn.expand("~/.cargo/bin/lean-ctx"),
  "/usr/local/bin/lean-ctx",
  "/opt/homebrew/bin/lean-ctx",
  vim.fn.expand("~/.local/bin/lean-ctx"),
}

function M.resolve()
  if cached_path then
    return cached_path
  end

  for _, candidate in ipairs(candidates) do
    if vim.fn.executable(candidate) == 1 then
      cached_path = candidate
      return candidate
    end
  end

  return nil
end

function M.run(args, callback)
  local bin = M.resolve()
  if not bin then
    callback("lean-ctx binary not found")
    return
  end

  local cmd = vim.list_extend({ bin }, args)
  local env = {
    LEAN_CTX_ACTIVE = "0",
    NO_COLOR = "1",
    PATH = vim.env.PATH,
    HOME = vim.env.HOME,
  }

  vim.system(cmd, {
    env = env,
    text = true,
  }, function(result)
    vim.schedule(function()
      local output = result.stdout or ""
      if output == "" then
        output = result.stderr or ""
      end
      callback(output)
    end)
  end)
end

return M

File: packages/vscode-lean-ctx/src/binary.ts
// Finding and running the `lean-ctx` binary. No `vscode` import.

import { execFile } from "child_process";
import { accessSync, constants } from "fs";
import * as path from "path";

/**
 * Where to look, in order: the configured path, then `PATH`, then the usual
 * install locations — an editor started from the Dock does not inherit the
 * login shell's `PATH`.
 */
export function candidatePaths(
  configured: string,
  envPath: string,
  home: string,
  platform: NodeJS.Platform,
): string[] {
  const exe = platform === "win32" ? "lean-ctx.exe" : "lean-ctx";
  const sep = platform === "win32" ? ";" : ":";
  const join = platform === "win32" ? path.win32.join : path.posix.join;
  const out: string[] = [];
  if (configured.trim()) out.push(configured.trim());
  for (const dir of envPath.split(sep)) if (dir) out.push(join(dir, exe));
  if (home) {
    out.push(join(home, ".local", "bin", exe), join(home, ".cargo", "bin", exe));
  }
  if (platform !== "win32") {
    out.push("/opt/homebrew/bin/lean-ctx", "/usr/local/bin/lean-ctx");
  }
  return [...new Set(out)];
}

function isExecutable(file: string): boolean {
  try {
    accessSync(file, process.platform === "win32" ? constants.F_OK : constants.X_OK);
    return true;
  } catch {
    return false;
  }
}

export function resolveBinary(configured: string): string | null {
  const candidates = candidatePaths(
    configured,
    process.env.PATH ?? "",
    process.env.HOME ?? process.env.USERPROFILE ?? "",
    process.platform,
  );
  return candidates.find(isExecutable) ?? null;
}

/** Runs the binary without a shell; stdout, or null on any failure. */
export function run(bin: string, args: string[], timeoutMs = 3000): Promise<string | null> {
  return new Promise((resolve) => {
    execFile(
      bin,
      args,
      { timeout: timeoutMs, env: { ...process.env, NO_COLOR: "1" }, maxBuffer: 256 * 1024 },
      (err, stdout) => resolve(err ? null : stdout),
    );
  });
}
